### PR TITLE
fix: revert mjs extension usage by default, make it an option

### LINKED ISSUE #9141: Project builds but fails on deploy
### Describe the bug

Today I updated my project with an NPM Update and it no longer works correctly. This is what my package.json look like:

The project builds succesfully:

And this is what happens when it gets deployed:

[Click here to see the issue for yourself](https://charbelnicolas.com/pockets)

### Reproduction

I could not reproduce this issue with a clean svelte-kit install. It is mostly a bug based on my project I guess... but this only happened after doing an NPM update today.

### Logs

_No response_

### System Info

```Shell
System:
    OS: Linux 6.1 Arch Linux
    CPU: (8) x64 Intel(R) Core(TM) i7-6700K CPU @ 4.00GHz
    Memory: 29.83 GB / 31.30 GB
    Container: Yes
    Shell: 5.1.16 - /bin/bash
  Binaries:
    Node: 19.6.1 - /usr/bin/node
    Yarn: 1.22.19 - /usr/bin/yarn
    npm: 8.19.2 - /usr/bin/npm
  Browsers:
    Chromium: 110.0.5481.100
```

### Severity

blocking all usage of SvelteKit

### Additional Information

_No response_

### PR DESCRIPTION
**Pull request by dummdidumm**
fixes #9141

hides the `.mjs` stuff behind an option which is `false` by default (else it's a breaking change, strictly speaking). I chose `output.mjs` as the option name. If we keep that, it may make sense to move `outDir` (top level currently) into that for version 2.

### Please don't delete this checklist! Before submitting the PR, please make sure you do the following:
- [x] It's really useful if your PR references an issue where it is discussed ahead of time. In many cases, features are absent for a reason. For large changes, please create an RFC: https://github.com/sveltejs/rfcs
- [x] This message body should clearly illustrate what problems it solves.
- [ ] Ideally, include a test that fails without this PR but passes with it.

### Tests
- [x] Run the tests with `pnpm test` and lint the project with `pnpm lint` and `pnpm check`

### Changesets
- [x] If your PR makes a change that should be noted in one or more packages' changelogs, generate a changeset by running `pnpm changeset` and following the prompts. Changesets that add features should be `minor` and those that fix bugs should be `patch`. Please prefix changeset messages with `feat:`, `fix:`, or `chore:`.

### CODE REVIEW

**Comment by Conduitry** [comment]
The purist in me wants two changesets for this, one (a fix) rolling back the previous change, and another (a feature) adding the flag.

File: packages/kit/src/exports/vite/index.js [hunk 1]
```diff
@@ -519,9 +519,14 @@ export function set_building() {
 								// we use .mjs for client-side modules, because this signals to Chrome (when it
 								// reads the <link rel="preload">) that it should parse the file as a module
 								// rather than as a script, preventing a double parse. Ideally we'd just use
-								// modulepreload, but Safari prevents that
-								entryFileNames: ssr ? '[name].js' : `${prefix}/[name].[hash].mjs`,
-								chunkFileNames: ssr ? 'chunks/[name].js' : `${prefix}/chunks/[name].[hash].mjs`,
+								// modulepreload, but Safari prevents that. We hide it behind an option because
+								// many CDN's don't have the proper MIME type for mjs files yet.
```
Comment (Conduitry) [inline packages/kit/src/exports/vite/index.js:523]: The first part of this code comment should be reworded as well. We don't want to say "we use .mjs for this reason, except actually we don't".

Comment (benmccann) [inline packages/kit/src/exports/vite/index.js:523]:
```suggestion
								// many CDNs don't have the proper MIME type for mjs files yet.
```

Comment (benmccann) [reply packages/kit/src/exports/vite/index.js:523]: I'd actually just remove this comment entirely because it's already documented where we define the option and this just duplicates that

**Comment by benmccann** [comment]
Do we need to add it to https://kit.svelte.dev/docs/configuration or is that auto-generated?

**Comment by dummdidumm** [comment]
Autogenerated

**Comment by Rich-Harris** [comment]
Rolling it back will hurt performance for the vast majority of cases where `.mjs` is supported. Should we add an option to use `.js` instead that defaults to `false`?

I recognise that there are semver arguments in favour of rolling it back, but I think it's worth considering impact.

**Comment by Rich-Harris** [comment]
> It's more important that it works than that it's fast

There's nuance here though. Suppose `.mjs` works for 99% of apps, and 15% of users are on Safari — that means that we're penalising those 15% of users, not for the sake of 1% of users but 1% of _developers_. 

I'm not wedded to one position or the other but the waterfall on mobile really does suck.

**Comment by benmccann** [comment]
Yeah. I don't know if `.mjs` works for 99% of apps though. It's already been reported broken for nginx, Azure, Go's Fiber server, and a corporate CDN

@dummdidumm can we also sneak in a fix to include `mjs` on this line? https://github.com/sveltejs/kit/issues/9141#issuecomment-1441424839

**Comment by benmccann** [comment]
Maybe we should see if we can remove the `.mjs` option and use an ETag alongside the expires header? (https://github.com/guybedford/es-module-shims/issues/354#issuecomment-1442571803)

**Comment by 42triangles** [comment]
@benmccann I don't think using anything related to headers would actually fix the original issue, considering it uses the static adapter which just outputs a bunch of files, with the headers then only being set by whichever server is serving them.

And in my opinion this very much is a breaking change for the static adapter for webservers that do not recognize the `.mjs` file extension, which in the original issue includes `nginx` (we're using [the Rust library `rocket`](https://github.com/SergioBenitez/Rocket), and it also [doesn't set the mimetype correctly](https://api.rocket.rs/v0.5-rc/rocket/http/struct.ContentType.html#method.from_extension), which for our use case (using another type that uses this internally) would require significant changes to support it or an update from `rocket` - and while `rocket` is most certainly not nearly as widely used, I do believe that this doesn't change the fact that it is a breaking change due to issues like these).

And as for performance - having an option you can turn on for using `.mjs` seems reasonable, or *at least* having the ability to turn it off if the issue is encountered (under the assumption that most bigger servers etc could handle it correctly).

**Comment by 42triangles** [comment]
Oh, then I misread that, sorry!

**Comment by GeoffreyBooth** [comment]
> I'm not wedded to one position or the other but the waterfall on mobile really does suck.

Wasn't this the case until 4-ish days ago? If it was okay until then, I would think it should be okay once again now, at least until a better solution is found. I understand the reluctance to roll back a performance gain, but it's not like the new speed has become so widely used in the last few days that users will react negatively to rolling it back.

> There’s nuance here though. Suppose `.mjs` works for 99% of apps, and 15% of users are on Safari — that means that we’re penalising those 15% of users, not for the sake of 1% of users but 1% of _developers_.

Unfortunately I don’t think it’s anywhere close to 99% of apps or users. It’s about the servers serving the files, and static webservers are updated on a very long timeline; [Apache 2.2 was end-of-life in 2018, but as of 2023-02-23 it’s still used by 3.5% of all websites](https://w3techs.com/technologies/details/ws-apache/2). Even [the _current_ version of Nginx still doesn’t serve .mjs with the correct Content-Type by default](https://trac.nginx.org/nginx/ticket/2216), nor does the current version of Apache httpd, even after all these years. Together, [Nginx and Apache account for 66.7% of all webservers](https://w3techs.com/technologies/overview/web_server).

I help manage modules for Node.js, and I’ve been studying this issue for a long time. In 2019 I made a test suite for checking the `Content-Type` provided for .mjs files by various common webservers: https://github.com/GeoffreyBooth/js-mjs-mime-type-test. Short version: even today, it’s still not safe to assume that .mjs will be served correctly by even the current versions of default-configured static webservers, to say nothing of older versions or CDNs.

I found https://github.com/sveltejs/kit/pull/9018#issuecomment-1432516596, and I must say, I’m very surprised to see that code in Chrome. It feels a lot like a spec violation. As far as I know, browsers aren’t supposed to read “file extensions” (really, the last part of the `pathname` of a URL) or use them for any purpose, especially not for determining types; that’s what the `Content-Type` header is for. Either that code is a bug, or maybe an intentional spec violation in the name of performance; but it doesn’t feel like something that should be relied upon. I understand the desire to avoid a waterfall, but I would suggest looking for a less suspect solution than relying on this questionable code.

File: packages/kit/src/exports/vite/index.js (hunk 1 above)
Comment (dummdidumm) [reply packages/kit/src/exports/vite/index.js:523]: reworded slightly

File: packages/kit/test/apps/basics/svelte.config.js [hunk 2]
```diff
@@ -1,6 +1,10 @@
 /** @type {import('@sveltejs/kit').Config} */
 const config = {
 	kit: {
+		output: {
+			preloadStrategy: 'preload-mjs'
```
Comment (dominikg) [inline packages/kit/test/apps/basics/svelte.config.js:5]: do we really want to switch our largest test app to a non-default strategy? i'd rather have a separate app test this

Comment (Rich-Harris) [reply packages/kit/test/apps/basics/svelte.config.js:5]: yeah, let's move it into `options`. the tests need updating anyway by the looks of things

File: packages/kit/src/runtime/server/page/render.js [hunk 3]
```diff
@@ -273,12 +273,13 @@ export async function render_response({
...
 
 		for (const path of included_modulepreloads) {
-			// we use modulepreload with the Link header for Chrome, along with
-			// <link rel="preload"> for Safari. This results in the fastest loading in
-			// the most used browsers, with no double-loading. Note that we need to use
-			// .mjs extensions for `preload` to behave like `modulepreload` in Chrome
+			// see the kit.output.preloadStrategy option for details on why we have multiple options here
 			link_header_preloads.add(`<${encodeURI(path)}>; rel="modulepreload"; nopush`);
-			head += `\n\t\t<link rel="preload" as="script" crossorigin="anonymous" href="${path}">`;
+			if (options.preload_strategy !== 'modulepreload') {
```
Comment (dominikg) [inline packages/kit/src/runtime/server/page/render.js:278]: hmm, is the rel=modulepreload link always added ? shouldn't it be either one or the other?

File: packages/kit/src/exports/vite/index.js [hunk 4]
```diff
@@ -516,12 +516,17 @@ export function set_building() {
...
-								// rather than as a script, preventing a double parse. Ideally we'd just use
-								// modulepreload, but Safari prevents that
-								entryFileNames: ssr ? '[name].js' : `${prefix}/[name].[hash].mjs`,
-								chunkFileNames: ssr ? 'chunks/[name].js' : `${prefix}/chunks/[name].[hash].mjs`,
+								// see the kit.output.preloadStrategy option for details on why we have multiple options here
+								entryFileNames: ssr
+									? '[name].js'
+									: `${prefix}/[name].[hash]${
+											kit.output.preloadStrategy === 'preload-mjs' ? '.mjs' : '.js'
+									  }`,
```
Comment (Rich-Harris) [inline packages/kit/src/exports/vite/index.js:524]: Can we DRY out this and the section below?

```suggestion
									: `${prefix}/[name].[hash].${ext}`,
```

File: packages/kit/types/index.d.ts [hunk 5]
```diff
@@ -422,6 +422,19 @@ export interface KitConfig {
...
+	 * Options related to the build output format
+	 */
+	output?: {
+		/**
+		 * What preload strategy to use for JavaScript files to avoid waterfalls:
+		 * - `modulepreload` - uses `<link rel="modulepreload">` to preload JavaScript files. Works only in Chromium-based browsers currently, and soon in Safari, too.
+		 * - `preload-js` - uses `<link rel="preload">` to preload JavaScript files. Works in all browsers but Firefox, and causes double-parsing of the script for Chromium-based browser.
+		 * - `preload-mjs` - uses `<link rel="preload">` to preload JavaScript files, but uses the `.mjs` extension. Works in Chromium-based browsers and Safari. Check of your provider/CDN has the correct MIME type for `.mjs` files before turning this on.
+		 * @default "modulepreload"
+		 */
```
Comment (Rich-Harris) [inline packages/kit/types/index.d.ts:435]: I think we can afford to add loads of context about these different options and what they entail

```suggestion
		/**
		 * SvelteKit will preload the JavaScript modules needed for the initial page to avoid import 'waterfalls', resulting in faster application startup. There
		 * are three strategies with different trade-offs:
		 * - `modulepreload` - uses `<link rel="modulepreload">`. This delivers the best results in Chromium-based browsers, but is currently ignored by Firefox and Safari (though support is coming to Safari soon).
		 * - `preload-js` - uses `<link rel="preload">`. Prevents waterfalls in Chromium and Safari, but Chromium will parse each module twice (once as a script, once as a module). Causes modules to be requested twice in Firefox. This is a good setting if you want to maximise performance for users on iOS devices at the cost of a very slight degradation for Chromium users.
		 * - `preload-mjs` - uses `<link rel="preload">` but with the `.mjs` extension which prevents double-parsing in Chromium. Some static webservers will fail to serve .mjs files with the correct MIME type, which will cause your application to break. If that doesn't apply to you, this is the option that will deliver the best performance for the largest number of users, until `modulepreload` is more widely supported.
		 * @default "modulepreload"
		 */
```

File: packages/kit/src/runtime/server/page/render.js (hunk 3 above)
Comment (dummdidumm) [reply packages/kit/src/runtime/server/page/render.js:278]: The modulepreload equivalent is the header links above. Since it's a header it doesn't hurt I think, but you're right we could put it inside an if block

Comment (Rich-Harris) [reply packages/kit/src/runtime/server/page/render.js:278]: Chrome doesn't respect `preload` headers, only the `<link>`. This might be a bug in Chrome, but I haven't got round to investigating and/or filing a ticket. Either way, I think we need to keep the `modulepreload` link headers for Chrome's benefit

Comment (dummdidumm) [reply packages/kit/src/runtime/server/page/render.js:278]: Wait, really? This means that chrome didn't get preload behavior previously, too, because that logic I implemented is basically the one we had prior to your PR which introduced the link preload. Mhmmm.
Isn't there also a related issue about the link headers being too large and crashing something? Do we need another "linkheaders" option?

File: packages/kit/types/index.d.ts [hunk 6]
```diff
@@ -427,10 +427,12 @@ export interface KitConfig {
 	 */
 	output?: {
 		/**
-		 * Whether or not to use the `.mjs` extension for JavaScript files. This makes it possible to preload JavaScript files in Safari, too (else only works in Chromium-based browsers).
-		 * Check if your provider has the correct MIME type for `.mjs` files before turning this on.
+		 * What preload strategy to use for JavaScript files to avoid waterfalls:
+		 * - `modulepreload` (default) - uses `<link rel="modulepreload">` to preload JavaScript files. Works only in Chromium-based browsers currently, and soon in Safari, too.
+		 * - `preload-js` - uses `<link rel="preload">` to preload JavaScript files. Works in all browsers but Firefox, and causes double-parsing of the script for Chromium-based browser.
+		 * - `preload-mjs` - uses `<link rel="preload">` to preload JavaScript files, but uses the `.mjs` extension. Works in Chromium-based browsers and Safari. Check of your provider/CDN has the correct MIME type for `.mjs` files before turning this on.
```
Comment (GeoffreyBooth) [inline packages/kit/types/index.d.ts:433]:
```suggestion
		 * - `preload-mjs` - uses `<link rel="preload">` to preload JavaScript files, but uses the `.mjs` extension. Works in Chromium-based browsers and Safari. Check if your server or CDN sets the correct `Content-Type` header for `.mjs` files (text/javascript) before turning this on; browsers will error on a missing or incorrect `Content-Type` header.
```

Thank you, this overall comment is much clearer.

File: packages/kit/test/apps/basics/svelte.config.js (hunk 2 above)
Comment (GeoffreyBooth) [reply packages/kit/test/apps/basics/svelte.config.js:5]: The recent bug reports around this could have been avoided if your test suite had a test like this:

1. Build the example app.
2. Download the latest `nginx` Docker image and run it, having it serve the build output.
3. Run headless Chrome and open the port served by the Nginx container.
4. Verify that there are no errors in the JavaScript console and that the page has minimal expected behavior (like clicking the button increments the counter).

The Docker orchestration parts could be copied from https://github.com/GeoffreyBooth/js-mjs-mime-type-test/blob/master/test.sh.

File: packages/kit/types/index.d.ts (hunk 5 above)
Comment (GeoffreyBooth) [reply packages/kit/types/index.d.ts:435]: Forgive me for joining this discussion late, but is there something I can read that explains the problem and the various solutions, and what you’re trying to achieve? Like are you trying to get the browser to download these other JavaScript files and parse them but not evaluate them?

For example, it occurs to me that you could just add a `script` block in the output HTML, like

```html
<script type="module">
import '_app/start.js' // Or whatever the generated full or root-prefixed path is for each chunk
</script>
```

But this would not only download but also evaluate each file (and its dependencies), which I assume you’re trying to avoid? You could also just `fetch` each one, to download and cache but not evaluate (though does a later `import` redownload?) or use dynamic `import()` to apply custom logic to determining which chunks to download and evaluate.

Basically I’m just wondering what your goal is and the various alternatives you tried, and what you found when you investigated each one. If that were written down somewhere (and maybe it already is), it would help others to try to improve on what you’ve implemented here. It would also help if there were a test in your suite to determine if what you’re trying to avoid, the “waterfalls,” happens or not.

File: packages/kit/test/apps/basics/svelte.config.js (hunk 2 above)
Comment (Rich-Harris) [reply packages/kit/test/apps/basics/svelte.config.js:5]: That's a huge amount of overhead to guard against a very unlikely regression

File: packages/kit/src/runtime/server/page/render.js (hunk 3 above)
Comment (Rich-Harris) [reply packages/kit/src/runtime/server/page/render.js:278]: Not sure I follow? we previously added `modulepreload` links to the HTTP header and still do (and should)

File: packages/kit/types/index.d.ts (hunk 5 above)
Comment (Rich-Harris) [reply packages/kit/types/index.d.ts:435]: The documentation in the comment above _is_ the explanation. We need to retain control of evaluation order because we can't run user code until some setup work has taken place — we _could_ refactor the code instead but it would increase bundle sizes, and could potentially box us in in future. `modulepreload` is the standard way to prevent import waterfalls.

`fetch`-based solutions don't work, unfortunately — things go into a different 'bucket' depending on whether they were fetched or imported.

File: packages/kit/test/apps/basics/svelte.config.js (hunk 2 above)
Comment (Rich-Harris) [reply packages/kit/test/apps/basics/svelte.config.js:5]: That's simply not going to happen. There are two things we can guard against:

* re-introducing `.mjs`, which would require a maintainer to ignore the 'this is why we don't use `.mjs`' comment that this PR adds
* every conceivable bug that could happen for every conceivable way someone could use SvelteKit, in which case our test suite would take infinite minutes to run, and we would spend all our lives maintaining the test suite inside of the actual code

Hindsight is 20/20 — it would have been nice not to have introduced this breaking change, but the reality of software development is that breakage happens. That's why we have versions!

**Review by Rich-Harris — APPROVED** [review]
